dhis-2/dhis-support/dhis-support-hibernate/src/main/java/org/hisp/dhis/hibernate/jsonb/type/JsonBinaryType.java

```diff
@@ -101,9 +107,36 @@
     }
 
     @Override
-    public boolean equals( Object x, Object y ) throws HibernateException
+    public boolean equals( Object x, Object y )
+            throws HibernateException
+    {
+        return x == y || (x != null && y != null && (x.equals( y ) || safeContentBasedEquals( x, y )));
+    }
+
+    private boolean safeContentBasedEquals( Object x, Object y )
+    {
+        Optional<Map<String, Object>> first = safelyConvertToMap( x );
+        Optional<Map<String, Object>> second = safelyConvertToMap( y );
+
+        if ( first.isPresent() && second.isPresent() )
+        {
+            return first.equals( second );
+        }
+
+        return false;
+    }
+
+    private Optional<Map<String, Object>> safelyConvertToMap( Object o )
     {
-        return x == y || !(x == null || y == null) && x.equals( y );
+        try
+        {
+            return Optional.ofNullable( resultingMapper.readValue( resultingMapper.writeValueAsString( o ),
```

Comment:
Is this Optional.ofNullable really needed?
The object is never null, can resultingMapper.readValue( resultingMapper.writeValueAsString( o ) be null?
If not, it is more expressive to use Optional.of

Reply:
yes, I'll switch to Optional.of()